File: src/main/java/com/jeffreyorazulike/simpletron/display/FileDisplay.java
/*
 * Copyright (C) 2021 Jeffrey Orazulike
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package com.jeffreyorazulike.simpletron.display;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

/**
 *
 * @author Jeffrey Orazulike
 * @apiNote Defines the way a message is displayed to a file
 */
public class FileDisplay implements Display {

    private final BufferedWriter writer;

    public FileDisplay(String path) throws IOException {
        writer = new BufferedWriter(new FileWriter(path));
    }

    @Override
    public void show(String message, Object... args) {
        try {
            writer.write(String.format(message, args));
            writer.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void close() {
        try {
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: core/impl/src/main/kotlin/com/jeffreyorazulike/simpletron/core/impl/component/FileDisplay.kt
package com.jeffreyorazulike.simpletron.core.impl.component

import com.jeffreyorazulike.simpletron.core.component.Display
import java.io.BufferedWriter
import java.io.FileWriter
import java.io.IOException

/**
 *
 * @author Jeffrey Orazulike [chukwudumebiorazulike@gmail.com]
 * Created on 22 at 8:11 PM
 *
 *
 * @param path the path to the file to open
 * @throws IOException if an exception occurred while trying to open the
 * file
 */
class FileDisplay(path: String) : Display() {
    override var isClosed = false
    private val writer: BufferedWriter

    init {
        writer = BufferedWriter(FileWriter(path))
    }

    @Throws(IllegalStateException::class, IOException::class)
    override fun show(message: String?) {
        check(!isClosed) { closedMessage }
        message?.let { writer.write(it) }
        writer.flush()
    }

    override fun close() {
        // throws illegal state exception if the component is already closed
        check(!isClosed) { closedMessage }
        writer.close()
        isClosed = true
    }
}
